mu4/appshell/qml/NotationPage/NotationToolBar.qml

```diff
@@ -1,49 +1,102 @@
-import QtQuick 2.7
-import QtQuick.Controls 2.2
+import QtQuick 2.12
+import QtQuick.Controls 2.12
+
 import MuseScore.NotationScene 1.0
 import MuseScore.UiComponents 1.0
+import MuseScore.Ui 1.0
 
 Rectangle {
     id: root
 
-    Row {
+    property alias orientation: gridView.orientation
+
+    QtObject {
+        id: privatesProperties
+
+        property bool isHorizontal: orientation === Qt.Horizontal
+    }
+
+    GridViewSectional {
+        id: gridView
         anchors.fill: parent
 
-        Repeater {
-            anchors.fill: parent
-            model: toolModel
-            delegate: ToolButton {
-                id: control
-                text: titleRole
-                enabled: enabledRole
-                down: checkedRole
-                onClicked: toolModel.click(nameRole)
-
-                contentItem: StyledTextLabel {
-                    text: control.text
-                    font: control.font
-                    opacity: enabled ? 1.0 : 0.5
-                    horizontalAlignment: Text.AlignHCenter
-                    verticalAlignment: Text.AlignVCenter
-                    elide: Text.ElideRight
-                }
-
-                background: Rectangle {
-                    implicitWidth: 40
-                    implicitHeight: 40
-                    color: control.enabled && (control.checked || control.highlighted) ? ui.theme.accentColor : ui.theme.buttonColor
-                    opacity: enabled ? ui.theme.accentOpacityNormal : ui.theme.itemOpacityDisabled
-                    visible: control.down || (control.enabled && (control.checked || control.highlighted))
-                }
+        sectionRole: "sectionRole"
+
+        cellWidth: 36
+        cellHeight: cellWidth
+
+        model: toolModel
+
+        sectionDelegate: Rectangle {
+            color: ui.theme.strokeColor
+        }
+
+        itemDelegate: FlatButton {
+            property var item: Boolean(itemModel) ? itemModel : null
+
+            normalStateColor: Boolean(item) && item.checkedRole ? ui.theme.accentColor : "transparent"
+
+            icon: Boolean(item) ? item.iconRole : null
+
+            onClicked: {
+                toolModel.click(item.nameRole)
             }
         }
     }
 
+    FlatButton {
+        id: customizeButton
+
+        anchors.margins: 8
+
+        icon: IconCode.CONFIGURE
+        normalStateColor: "transparent"
+
+        onClicked: {
+            api.launcher.open("musescore://notation/noteinputbarcustomise")
+        }
+    }
+
     NotationToolBarModel {
         id: toolModel
     }
 
     Component.onCompleted: {
         toolModel.load()
     }
+
+    states: [
+        State {
+            when: privatesProperties.isHorizontal
+            PropertyChanges {
+                target: gridView
+                sectionWidth: 2
+                sectionHeight: root.height
+                rows: 1
+                columns: -1
```

Comment:
Why not 0?

Reply:
This means there is no column limit. View can have as many columns as possible.

Reply:
Can you add readonly property like NO_LIMIT instead of this magic number?